# Stack Overflow Public Dataset

This dataset, hosted on **Google BigQuery** under the `bigquery-public-data` project, contains a rich collection of data from **Stack Overflow**, the popular Q&A platform for developers. It includes information about questions, answers, users, comments, and votes, spanning many years of community activity.

The dataset is organized in several tables, allowing analysis of user behavior, trends in programming languages, popular topics, and community dynamics.

In [8]:
from google.cloud import bigquery

# Construct a BigQuery client object.
client = bigquery.Client()

# construct a reference to the "stackoverflow" dataset
dataset_ref = client.dataset("stackoverflow", project="bigquery-public-data")

# API request - fetch the dataset
dataset = client.get_dataset(dataset_ref)

/home/jd2208/anaconda3/envs/datanalysis/lib/python3.13/site-packages/google/auth/_default.py:76: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK without a quota project. You might receive a "quota exceeded" or "API not enabled" error. See the following page for troubleshooting: https://cloud.google.com/docs/authentication/adc-troubleshooting/user-creds. 
  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)
/home/jd2208/anaconda3/envs/datanalysis/lib/python3.13/site-packages/google/auth/_default.py:76: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK without a quota project. You might receive a "quota exceeded" or "API not enabled" error. See the following page for troubleshooting: https://cloud.google.com/docs/authentication/adc-troubleshooting/user-creds. 
  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)


### 1) Explore the data

Before writing queries, We'll want to see what tables are available.

In [9]:
# Get tables available in the dataset
tables = list(client.list_tables(dataset))

# Print the list of tables
list_of_tables = [table.table_id for table in tables]
print("Tables in dataset {}: {}".format(dataset.dataset_id, list_of_tables))

Tables in dataset stackoverflow: ['badges', 'comments', 'post_history', 'post_links', 'posts_answers', 'posts_moderator_nomination', 'posts_orphaned_tag_wiki', 'posts_privilege_wiki', 'posts_questions', 'posts_tag_wiki', 'posts_tag_wiki_excerpt', 'posts_wiki_placeholder', 'stackoverflow_posts', 'tags', 'users', 'votes']


In [10]:
# construct a reference to the "posts_answers" table
answer_table_ref = dataset_ref.table("posts_answers")

# API request - fetch the table
answers_table = client.get_table(answer_table_ref)

# Print information on the "posts_answers" table
client.list_rows(answers_table, max_results=5).to_dataframe()

id title                                               body  \
0    18  None  <p>For a table like this:</p>\n\n<pre><code>CR...   
1   165  None  <p>You can use a <a href="http://sharpdevelop....   
2  1028  None  <p>The VB code looks something like this:</p>\...   
3  1073  None  <p>My first choice would be a dedicated heap t...   
4  1260  None  <p>I found the answer. all you have to do is a...   

  accepted_answer_id answer_count  comment_count community_owned_date  \
0               None         None              2                  NaT   
1               None         None              0                  NaT   
2               None         None              0                  NaT   
3               None         None              0                  NaT   
4               None         None              0                  NaT   

                     creation_date favorite_count  \
0 2008-08-01 05:12:44.193000+00:00           None   
1 2008-08-01 18:04:25.023000+00:00           None   
2 2008-08-04 04:58:40.300000+00:00           None   
3 2008-08-04 07:51:02.997000+00:00           None   
4 2008-08-04 14:06:02.863000+00:00           None   

                last_activity_date                   last_edit_date  \
0 2016-06-02 05:56:26.060000+00:00 2016-06-02 05:56:26.060000+00:00   
1 2019-04-06 14:03:51.080000+00:00 2019-04-06 14:03:51.080000+00:00   
2 2013-02-07 13:22:14.680000+00:00 2013-02-07 13:22:14.680000+00:00   
3 2015-09-01 17:32:32.120000+00:00 2015-09-01 17:32:32.120000+00:00   
4 2016-12-20 08:38:48.867000+00:00 2016-12-20 08:38:48.867000+00:00   

  last_editor_display_name  last_editor_user_id owner_display_name  \
0              Jeff Atwood               126039             phpguy   
1                     None              1721793        user2189331   
2                     None               395659        user2189331   
3                     None                45459        user2189331   
4                     None              1221571                Jin   

   owner_user_id  parent_id  post_type_id  score  tags view_count  
0           <NA>         17             2     59  None       None  
1           <NA>        145             2     10  None       None  
2           <NA>        947             2      8  None       None  
3           <NA>       1069             2     29  None       None  
4           <NA>       1229             2      1  None       None

In [11]:
# construc a reference to the "posts_questions" table
question_table_ref = dataset_ref.table("posts_questions")

# API request - fetch the table
question_table = client.get_table(question_table_ref)

# Print information on the "posts_questions" table
client.list_rows(question_table, max_results=5).to_dataframe()

id                                              title  \
0  320268          Html.ActionLink doesn’t render # properly   
1  324003                                Primitive recursion   
2  390605                                 While vs. Do While   
3  413246                        Protect ASP.NET Source code   
4  454921  Difference between "int[] myArray" and "int my...   

                                                body  accepted_answer_id  \
0  <p>When using Html.ActionLink passing a string...                <NA>   
1  <p>how  will i define the function 'simplify' ...                <NA>   
2  <p>I've seen both the blocks of code in use se...              390608   
3  <p>Im currently doing some research in how to ...                <NA>   
4  <blockquote>\n  <p><strong>Possible Duplicate:...              454928   

   answer_count  comment_count community_owned_date  \
0             0              0                  NaT   
1             0              0                  NaT   
2             0              0                  NaT   
3             0              0                  NaT   
4             0              0                  NaT   

                     creation_date  favorite_count  \
0 2008-11-26 10:42:37.477000+00:00               0   
1 2008-11-27 15:12:37.497000+00:00               0   
2 2008-12-24 01:49:54.230000+00:00               2   
3 2009-01-05 14:23:51.040000+00:00               0   
4 2009-01-18 10:22:52.177000+00:00               0   

                last_activity_date                   last_edit_date  \
0 2009-02-06 20:13:54.370000+00:00                              NaT   
1 2012-09-25 19:54:40.597000+00:00 2012-09-25 19:54:40.597000+00:00   
2 2008-12-24 03:08:55.897000+00:00                              NaT   
3 2009-03-24 21:30:22.370000+00:00 2009-01-05 14:42:28.257000+00:00   
4 2009-01-18 10:30:50.930000+00:00 2017-05-23 11:49:26.567000+00:00   

  last_editor_display_name  last_editor_user_id owner_display_name  \
0                     None                 <NA>              Paulo   
1                   Marcin                 1288               None   
2                     None                 <NA>          Unkwntech   
3             Tom Anderson                13502            Velnias   
4                     None                   -1       Evan Fosmark   

   owner_user_id parent_id  post_type_id  score  \
0           <NA>      None             1      0   
1          41000      None             1      0   
2            115      None             1      0   
3           <NA>      None             1      0   
4          49701      None             1      0   

                                                tags  view_count  
0                                        asp.net-mvc         390  
1  haskell|lambda|functional-programming|lambda-c...         497  
2                            language-agnostic|loops       11262  
3                     asp.net|deployment|obfuscation        4823  
4                                        java|arrays         798

If we are looking for people interested in a specific topic, these tables, "post_answers" and "post_questions," are the right place. 

## Finding Users Interested in a Specific Topic 

To identify people interested in a specific topic on Stack Overflow, we can use the relationship between `post_questions` and `post_answers`. The relationshibs parent_id and owner_user_id allow us to do is. For example the questions about bigquery.  


In [27]:
# query to find questions and answers related to bigquery
answers_query = """
     SELECT q.id, a.parent_id, a.body, a.owner_user_id
     FROM `bigquery-public-data.stackoverflow.posts_questions` AS q
     INNER JOIN `bigquery-public-data.stackoverflow.posts_answers` AS a
     ON q.id = a.parent_id
     WHERE q.tags LIKE '%bigquery%'
     """

# Run the query and convert the results to a pandas DataFrame
answers_query_job = client.query(answers_query)

# API request - fetch the results
answers_df = answers_query_job.to_dataframe()

answers_df.head()

id  parent_id                                               body  \
0  55849940   55849940  <p>You have a BQ table with column attribute_v...   
1  55855557   55855557  <p>You may need to pre-process your input data...   
2  55867453   55867453  <p>This is an example of your SQL code with sm...   
3  55868054   55868054  <p>Try this,</p>\n\n<pre><code>select * from x...   
4  55861179   55861179  <p>There is no way to do this directly.  Inste...   

   owner_user_id  
0       11403457  
1       11403457  
2        1031958  
3        7462954  
4        7644371

a.parent_id which identifies the q.id of the question each answer is responding to. We have therefore found the questions that have bigquery somewhere in the tag.

### Exploratory Question

We want to identify users who have answered **many questions** related to the tag `"bigquery"`.

In [31]:
# query to find expert users who have provided answers related to bigquery
expert_users_query = """
                   SELECT a.owner_user_id as user_id, COUNT(1) AS numbers_of_answers
                   FROM `bigquery-public-data.stackoverflow.posts_questions` AS q
                   INNER JOIN `bigquery-public-data.stackoverflow.posts_answers` AS a
                   ON q.id = a.parent_id
                   WHERE q.tags LIKE '%bigquery%'
                   GROUP BY a.owner_user_id
                   ORDER BY numbers_of_answers DESC
                   """

# set up the query
bigquery_expert_users_query = client.query(expert_users_query)

# API request - run the query and return panda Dataframe
bigquery_expert_result = bigquery_expert_users_query.to_dataframe()

bigquery_expert_result.head()



user_id  numbers_of_answers
0  5221944                5203
1  1144035                1634
2   132438                 898
3  6253347                 737
4  1366527                 620

So the user of with ID 5221944 has participate of 5203 discussion about bigquery and therefore should be expert person in bigquery. 